File: python/tests/test_euler004.py
import sys
import os
import unittest

sys.path.insert(
    0,
    os.path.abspath(os.path.join(os.path.dirname(__file__), "..")),
)
from solutions.euler004 import (  # noqa: E402
    is_palindrome,
    largest_palindrome_product,
)


class TestEuler004(unittest.TestCase):
    def test_is_palindrome(self):
        """Test the is_palindrome function with various inputs."""
        # Test single digit numbers
        self.assertTrue(is_palindrome(1))
        self.assertTrue(is_palindrome(9))

        # Test two digit numbers
        self.assertTrue(is_palindrome(11))
        self.assertTrue(is_palindrome(99))
        self.assertFalse(is_palindrome(12))
        self.assertFalse(is_palindrome(98))

        # Test three digit numbers
        self.assertTrue(is_palindrome(121))
        self.assertTrue(is_palindrome(999))
        self.assertFalse(is_palindrome(123))
        self.assertFalse(is_palindrome(998))

        # Test larger numbers
        self.assertTrue(is_palindrome(9009))
        self.assertTrue(is_palindrome(12321))
        self.assertFalse(is_palindrome(12345))

    def test_largest_palindrome_product(self):
        """Test the largest_palindrome_product function."""
        result = largest_palindrome_product()
        self.assertEqual(result, 906609)  # The actual answer

        # Verify that the result is a palindrome
        self.assertTrue(is_palindrome(result))

        # Verify that the result is a product of two 3-digit numbers
        factors_found = False
        for i in range(999, 99, -1):
            if result % i == 0:
                other_factor = result // i
                if 100 <= other_factor <= 999:
                    factors_found = True
                    break
        self.assertTrue(factors_found)


if __name__ == "__main__":
    unittest.main()

File: ruby/tests/test_euler004.rb
# frozen_string_literal: true

# !/usr/bin/env ruby

require 'test/unit'
require_relative '../euler004'

# Test cases for Project Euler Problem 4: Largest Palindrome Product
class TestEuler004 < Test::Unit::TestCase
  def test_single_digit_palindromes
    assert(palindrome?(1))
    assert(palindrome?(9))
  end

  def test_two_digit_palindromes
    assert(palindrome?(11))
    assert(palindrome?(99))
    assert_false(palindrome?(12))
    assert_false(palindrome?(98))
  end

  def test_three_digit_palindromes
    assert(palindrome?(121))
    assert(palindrome?(999))
    assert_false(palindrome?(123))
    assert_false(palindrome?(998))
  end

  def test_larger_palindromes
    assert(palindrome?(9009))
    assert(palindrome?(12_321))
    assert_false(palindrome?(12_345))
  end

  def test_largest_palindrome_product
    result = largest_palindrome_product
    assert_equal(906_609, result) # The actual answer

    # Verify that the result is a palindrome
    assert(palindrome?(result))

    # Verify that the result is a product of two 3-digit numbers
    factors_found = false
    999.downto(100) do |i|
      next unless (result % i).zero?

      other_factor = result / i
      if other_factor >= 100 && other_factor <= 999
        factors_found = true
        break
      end
    end
    assert(factors_found)
  end
end
